jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/compress/ByteAccumulator.java

```diff
@@ -19,56 +19,77 @@
 package org.eclipse.jetty.websocket.common.extensions.compress;
 
 import java.nio.ByteBuffer;
-import java.util.ArrayList;
-import java.util.List;
 
+import org.eclipse.jetty.io.ByteBufferAccumulator;
+import org.eclipse.jetty.io.ByteBufferPool;
 import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.websocket.api.MessageTooLargeException;
 
-public class ByteAccumulator
+public class ByteAccumulator implements AutoCloseable
 {
-    private final List<byte[]> chunks = new ArrayList<>();
+    private final ByteBufferAccumulator accumulator;
     private final int maxSize;
     private int length = 0;
 
     public ByteAccumulator(int maxOverallBufferSize)
     {
-        this.maxSize = maxOverallBufferSize;
+        this(maxOverallBufferSize, null);
     }
 
-    public void copyChunk(byte[] buf, int offset, int length)
+    public ByteAccumulator(int maxOverallBufferSize, ByteBufferPool byteBufferPool)
     {
-        if (this.length + length > maxSize)
-        {
-            String err = String.format("Resulting message size [%,d] is too large for configured max of [%,d]", this.length + length, maxSize);
-            throw new MessageTooLargeException(err);
-        }
+        this.maxSize = maxOverallBufferSize;
+        this.accumulator = new ByteBufferAccumulator(byteBufferPool);
+    }
 
-        byte[] copy = new byte[length - offset];
-        System.arraycopy(buf, offset, copy, 0, length);
+    public int getLength()
+    {
+        return accumulator.getLength();
+    }
 
-        chunks.add(copy);
-        this.length += length;
+    public ByteBuffer getBuffer(int minAllocationSize)
+    {
+        return accumulator.getBuffer(minAllocationSize);
     }
 
-    public int getLength()
+    public void copyChunk(byte[] buf, int offset, int length)
```

Comment:
Are the copy methods actually used now?   Isn't the accumulator filled by calling `getBuffer` and writing directly to that?   
If we don't need the copy methods then deprecate or remove these methods.

Alternately leave `ByteAccumulator` as is and deprecate the whole class.  Instead directly use `ByteBufferAccumulator`

Reply:
I think we need to maintain the `ByteAccumulator` to not break anyone. I have tried to only add new methods and not change the signatures of existing ones. So these copy methods stay.

I will deprecate the whole class.

Reply:
Then I'd just leave the entire implementation as it was - and we don't use it.

Reply:
We have to use it because it is used in method signatures.
```
protected ByteAccumulator newByteAccumulator();
protected void forwardIncoming(Frame frame, ByteAccumulator accumulator);
protected void decompress(ByteAccumulator accumulator, ByteBuffer buf);
```

I think this is the best we can do, it still works if you use it the original way, but it has additional methods to use it more efficiently.